sdk/typescript/src/types/sui-bcs.ts

```diff
@@ -352,12 +265,190 @@ export function deserializeTransactionBytesToTransactionData(
+    struct: {
+      recipient: 'address',
+      object_ref: 'SuiObjectRef',
+    },
+  },
+
+  PayTx: {
+    struct: {
+      coins: 'vector<SuiObjectRef>',
+      recipients: 'vector<address>',
+      amounts: 'vector<u64>',
+    },
+  },
+
+  PaySuiTx: {
+    struct: {
+      coins: 'vector<SuiObjectRef>',
+      recipients: 'vector<address>',
+      amounts: 'vector<u64>',
+    },
+  },
+
+  PayAllSuiTx: {
+    struct: {
+      coins: 'vector<SuiObjectRef>',
+      recipient: 'address',
+    },
+  },
+
+  TransferSuiTx: {
+    struct: {
+      recipient: 'address',
+      amount: 'Option<u64>',
+    },
+  },
+
+  PublishTx: {
+    struct: {
+      modules: 'vector<vector<u8>>',
+    },
+  },
+
+  SharedObjectRef: {
+    struct: {
+      objectId: 'address',
+      initialSharedVersion: 'u64',
+    },
+  },
+
+  ObjectArg: {
+    enum: {
+      ImmOrOwned: 'SuiObjectRef',
+      Shared: 'SharedObjectRef',
+    },
+  },
+
+  CallArg: {
+    enum: {
+      Pure: 'vector<u8>',
+      Object: 'ObjectArg',
+      ObjVec: 'vector<ObjectArg>',
+    }
+  },
+
+  TypeTag: {
+    enum: {
+      bool: null,
+      u8: null,
+      u64: null,
+      u128: null,
+      address: null,
+      signer: null,
+      vector: 'TypeTag',
+      struct: 'StructTag',
+      u16: null,
+      u32: null,
+      u256: null,
+    },
+  },
+
+  StructTag: {
+    struct: {
+      address: 'address',
+      module: 'string',
+      name: 'string',
+      typeParams: 'vector<TypeTag>',
+    },
+  },
+
+  MoveCallTx: {
+    struct: {
+      package: 'address',
+      module: 'string',
+      function: 'string',
+      typeArguments: 'vector<TypeTag>',
+      arguments: 'vector<CallArg>',
+    },
+  },
+
+  Transaction: {
+    enum: {
+      TransferObject: 'TransferObjectTx',
+      Publish: 'PublishTx',
+      Call: 'MoveCallTx',
+      TransferSui: 'TransferSuiTx',
+      Pay: 'PayTx',
+      PaySui: 'PaySuiTx',
+      PayAllSui: 'PayAllSuiTx',
+    },
+  },
+
+  TransactionKind: {
+    enum: {
+      Single: 'Transaction',
+      Batch: 'vector<Transaction>',
+    },
+  },
+
+  TransactionData: {
+    struct: {
+      kind: 'TransactionKind',
+      sender: 'address',
+      gasPayment: 'SuiObjectRef',
+      gasPrice: 'u64',
+      gasBudget: 'u64',
+    },
+  },
+
+  // Signed transaction data needed to generate transaction digest.
+  SenderSignedData: {
+    struct: {
+      data: 'TransactionData',
+      txSignature: 'vector<u8>',
+    },
+  },
+};
+
+const BCS_0_23_SPEC = {
+  ...BCS_SPEC,
+  MoveCallTx: {
+    struct: {
+      package: 'SuiObjectRef',
+      module: 'string',
+      function: 'string',
+      typeArguments: 'vector<TypeTag>',
+      arguments: 'vector<CallArg>',
+    },
+  },
+};
+
+const bcs = new BCS(getSuiMoveConfig());
+registerUTF8String(bcs);
+registerObjectDigest(bcs);
+registerTypes(bcs, BCS_SPEC);
+
+// ========== Backward Compatibility (remove after v0.24 deploys) ===========
+const bcs_0_23 = new BCS(getSuiMoveConfig());
```

Comment:
@damirka, heads up, I tried to implement the two versions of the BCS schema using the copy constructor on `BCS` but unfortunately it doesn't quite work, because the `encode` and `decode` closures inside the `types` map capture the instance of `BCS` that they were created for.

This means that in an example like this:

```js
const a = BCS();
a.registerStructType('Foo', { 'bar': 'string' })
a.registerStructType('Baz', { 'foo': 'Foo' })

const b = BCS(a);
b.registerStructType('Foo', { 'bar': 'u64' });

b.ser('Baz', { foo: { bar: 42 } });
```

The serialization fails because the serializer for `Baz` in `b` is still holding on to `a`, so when it looks up the type definition for `Foo`, it gets `a`'s.